smac/facade/hydra_facade.py

```diff
@@ -33,56 +86,115 @@ class Hydra(object):
+            valX => train_set * 100/X
+        incs_per_round: int
+            Number of incumbents to keep per round
+        n_optimizers: int
+            Number of optimizers to run in parallel per round
+        rng: int/np.random.RandomState
+            The randomState/seed to pass to each smac run
+        run_id: int
+            run_id for this hydra run
+        tae: ExecuteTARun
+            Target Algorithm Runner (supports old and aclib format)
 
+        """
         self.logger = logging.getLogger(
             self.__module__ + "." + self.__class__.__name__)
 
         self.n_iterations = n_iterations
-
-        self.logger = logging.getLogger(
-            self.__module__ + "." + self.__class__.__name__)
-
         self.scenario = scenario
-        self.run_id = run_id
+        self.run_id, self.rng = self._get_rng(rng, run_id)
         self.kwargs = kwargs
         self.output_dir = None
         self.top_dir = None
         self.solver = None
+        self.portfolio = None
         self.rh = RunHistory(average_cost)
+        self._tae = tae
         self.tae = tae(ta=self.scenario.ta, run_obj=self.scenario.run_obj)
-        self.tae_type = tae
+        if incs_per_round <= 0:
+            self.logger.warning('Invalid value in %s: %d. Setting to 1', 'incs_per_round', incs_per_round)
+        self.incs_per_round = max(incs_per_round, 1)
+        if n_optimizers <= 0:
+            self.logger.warning('Invalid value in %s: %d. Setting to 1', 'n_optimizers', n_optimizers)
+        self.n_optimizers = max(n_optimizers, 1)
+        self.val_set = self._get_validation_set(val_set)
+        self.cost_per_inst = {}
+
+    def _get_validation_set(self, val_set: str, delete: bool=True) -> typing.List[str]:
+        """
+        Create small validation set for hydra to determine incumbent performance
 
-    def optimize(self):
-        """Optimizes the algorithm provided in scenario (given in constructor)
+        Parameters
+        ----------
+        val_set: str
+            Set to validate incumbent(s) on. [train, valX].
+            train => whole training set,
+            valX => train_set * 100/X
+        delete: bool
+            Flag to delete all validation instances from the training set
+
+        """
+        if val_set == 'train':
+            return self.scenario.train_insts
+        elif val_set[:3] != 'val':
+            self.logger.warning('Can not determine validation set size. Using full training-set!')
+            return self.scenario.train_insts
+        else:
+            size = int(val_set[3:])/100
+            if size < 0 or size > 1:
+                raise ValueError('X too large in valX')
+            insts = np.array(self.scenario.train_insts)
+            # just to make sure this also works with the small example we have to round up to 3
+            size = max(np.floor(insts.shape[0] * size).astype(int), 3)
+            ids = np.random.choice(insts.shape[0], size, replace=False)
```

Comment:
why do you use `np.random.choice`? Since these are strings, `random.choice` should also do the trick.

Reply:
ids are actually ints and `random.choice` would only return one value.
If I were to use `random.choices` it would choose *with replacement* (see https://docs.python.org/3/library/random.html). As we don't want to have indices for the validation set that contains duplicates `np.random.choice(upper_limit, number_of_choices, replace=False)` is the easiest way to do it.

`random.sample` would sample without replacement but wouldn't give me the IDs, which I use to remove the validation data from the training set.